### PR TITLE
[#57] Interceptors implementation

### PR DESCRIPTION
**Pull request by RahimovIR**

### CODE REVIEW

File: src/Pulsar.Client/Internal/ConsumerImpl.fs [hunk 1]
```diff
@@ -776,45 +777,55 @@ type internal ConsumerImpl internal (consumerConfig: ConsumerConfiguration, clie
...
+                throwIfNotNull exn
+                
                 mb.Post(Acknowledge(msgId, AckType.Individual))
             }
             
         member this.AcknowledgeAsync (msgs: Messages) =
             task {
-                connectionHandler.CheckIfActive()
+                for msg in msgs do
+                    interceptors.OnAcknowledge(this, msg.MessageId, null)
```
Comment (Lanayx) [inline src/Pulsar.Client/Internal/ConsumerImpl.fs:790]: Why is null there?

File: src/Pulsar.Client/Internal/ConsumerImpl.fs [hunk 2]
```diff
@@ -834,13 +845,19 @@ type internal ConsumerImpl internal (consumerConfig: ConsumerConfiguration, clie
...
+                throwIfNotNull exn
+
                 mb.Post(NegativeAcknowledge(msgId))
             }
 
         member this.NegativeAcknowledge (msgs: Messages)  =
             task {
-                connectionHandler.CheckIfActive()
+                for msg in msgs do
+                    interceptors.OnNegativeAcksSend(this, msg.MessageId, null)
```
Comment (Lanayx) [inline src/Pulsar.Client/Internal/ConsumerImpl.fs:858]: Why is null there?

File: src/Pulsar.Client/Internal/ConsumerImpl.fs (hunk 1 above)
Comment (RahimovIR) [reply src/Pulsar.Client/Internal/ConsumerImpl.fs:790]: You suggested removing the check, maybe I misunderstood
https://github.com/RahimovIR/pulsar-client-dotnet/commit/d459f9526240adf4a7764efff31f1b73b8c7b0b9#r37607525

File: src/Pulsar.Client/Api/IConsumerInterceptor.fs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    
+    abstract member Close: unit -> unit
+    
+    abstract member BeforeConsume: consumer:IConsumer * message:Message -> Message  
+
+    abstract member OnAcknowledge: consumer:IConsumer * messageId:MessageId * ``exception``:Exception -> unit
+
+    abstract member OnAcknowledgeCumulative: consumer:IConsumer * messageId:MessageId * ``exception``:Exception -> unit
+    
+    abstract member OnNegativeAcksSend: consumer:IConsumer * messageId:MessageId * ``exception``:Exception -> unit
```
Comment (Lanayx) [inline src/Pulsar.Client/Api/IConsumerInterceptor.fs:16]: OnNegativeAcksSend doesn't accept exception, check Java version. It should accept `HashSet<MessageId>`

File: src/Pulsar.Client/Api/IConsumerInterceptor.fs [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    
+    abstract member OnAckTimeoutSend: consumer:IConsumer * messageId:MessageId * ``exception``:Exception -> unit
```
Comment (Lanayx) [inline src/Pulsar.Client/Api/IConsumerInterceptor.fs:18]: OnAckTimeoutSend doesn't accept exception, check Java version. It should accept `HashSet<MessageId>`

**Pull request closed without merging**